metric-example/query-by-example.md

```diff
@@ -0,0 +1,516 @@
+```
+That's better. You can see that by combining transformations, we can design powerful queries and slice and dice the data as we need it.
+
+# Scenario 9: Average Session Duration of Non-robot First-time Users with Advanced Transformer Chains
+
+>**Task**
+>We need to provide data on average session duration, but filter out some unwanted types of visits which do not represent a real user.
+
+With what we have learned up until now, no metric seems to be a perfect match for the following use case: We want to calculate the average session duration of first-time users, regardless of whether the user is synthetic or a real user. Some metrics come close, but are not perfect. Luckily, transformer chains allow us to derive new metrics from existing ones that exactly meet our needs.
+
+Consider for example `builtin:apps.web.sessionDuration`. Looking at response data we see that it does give us the session duration, and whether the visitor was new, but also reports recurring users and robots, cluttering our result with data that we don't need:
+```
+metricId,dt.entity.application,VisitorType,VisitType,time,value
+builtin:apps.web.sessionDuration,APPLICATION-DFD07AE5A6077853,RECURRING_VISITOR,REAL_USER,2019-04-09 13:31:00,7.0866E7
+builtin:apps.web.sessionDuration,APPLICATION-923D6049CBA84FBD,NEW_VISITOR,ROBOT,2019-04-09 13:31:00,4.61475E7
+...
+```
+Chaining multiple transformers together, we can derive a new metric from `builtin:apps.web.sessionDuration` that does exactly what we need. For this, we need `:filter` and `:merge` in our tool box.
+
+Examining the result data we see that additional dimensions tell us whether the numbers were recorded for new visitors or recurring ones and whether the visit was from a robot, a synthetic user or a real user. We are only interested in new visitors that are not robots. Whether they are synthetic or real is irrelevant for our use case.
+
+Dropping robots is easy with :`filter`, which drops results based on a condition. The condition may contain logical connectives and can perform checks on the values of dimension. Our expectations can be formalized as follows: `VisitorType` should be equal to `NEW_VISITOR`, while `VisitType` needs to be not equal to `ROBOT`. This can be expressed in terms of a filter, which we can append to our metric selector:
+```
+:filter(and(eq(VisitorType,NEW_VISITOR),ne(VisitType,ROBOT)))
+```
+We observe that the payload has changed to only contain the result data we are interested in. For each application, two data points remain per timeslot, one for `SYNTHETIC`, and one for `REAL_USER`, while the visitor type is constant at `NEW_VISITOR`. We actually want the average of synthetic and real, so that we have one data point per application. We can easily view them as a combined data point by using a `merge` transformer. It takes the names or indexes of dimensions to remove as arguments, merging data points that have become equal when the dimension is removed:

```

Comment:
```suggestion
We observe that the payload has changed to only contain the result data we are interested in. For each application, two data points remain per timeslot, one for `SYNTHETIC`, and one for `REAL_USER`, while the visitor type is constant at `NEW_VISITOR`. We actually want the average of synthetic and real, so that we have one data point per application. We can easily view them as a combined data point by using a `merge` transformer. It takes the names or indexes of dimensions to remove as arguments, merging data points that become equal when the dimensions are removed:

```